File: backend/api/confidence_dashboard.py
"""
Confidence Dashboard API endpoints.

This module provides endpoints for the AI Confidence Scoring & Transparency Dashboard,
including aggregated confidence metrics, confidence distribution analysis, uncertainty
indicators, and model accuracy insights.
"""
import csv
import io
import logging
from datetime import datetime
from enum import Enum
from typing import Any, Dict, Optional

from fastapi import APIRouter, Body, HTTPException, Query, status
from fastapi.responses import JSONResponse
from pydantic import BaseModel, Field

logger = logging.getLogger(__name__)

router = APIRouter()


# Enums
class ConfidenceExportFormat(str, Enum):
    """Supported export formats for confidence report data."""

    JSON = "json"
    CSV = "csv"


# Response Models
class ConfidenceDistribution(BaseModel):
    """Confidence score distribution across different ranges."""

    high_confidence: int = Field(..., description="Number of rankings with confidence >= 0.8")
    medium_confidence: int = Field(..., description="Number of rankings with confidence 0.5-0.8")
    low_confidence: int = Field(..., description="Number of rankings with confidence < 0.5")
    high_confidence_percentage: float = Field(..., description="Percentage of high confidence rankings")
    medium_confidence_percentage: float = Field(..., description="Percentage of medium confidence rankings")
    low_confidence_percentage: float = Field(..., description="Percentage of low confidence rankings")


class UncertaintyIndicators(BaseModel):
    """Indicators for model uncertainty and edge cases."""

    low_confidence_count: int = Field(..., description="Number of low confidence predictions (<0.5)")
    high_uncertainty_count: int = Field(..., description="Number of predictions with wide confidence intervals")
    average_confidence_interval_width: float = Field(..., description="Average width of confidence intervals")
    flagged_for_review: int = Field(..., description="Number of candidates flagged for manual review")


class ModelAccuracySummary(BaseModel):
    """Summary of model accuracy based on historical performance."""

    overall_accuracy: float = Field(..., description="Overall model accuracy (0-1)")
    precision: float = Field(..., description="Model precision (0-1)")
    recall: float = Field(..., description="Model recall (0-1)")
    f1_score: float = Field(..., description="Model F1 score (0-1)")
    total_predictions: int = Field(..., description="Total number of predictions made")
    feedback_count: int = Field(..., description="Number of predictions with recruiter feedback")


class ConfidenceMetricsResponse(BaseModel):
    """Response model for confidence dashboard metrics."""

    average_confidence: float = Field(..., description="Average confidence score across all rankings (0-1)")
    median_confidence: float = Field(..., description="Median confidence score (0-1)")
    min_confidence: float = Field(..., description="Minimum confidence score (0-1)")
    max_confidence: float = Field(..., description="Maximum confidence score (0-1)")
    total_rankings: int = Field(..., description="Total number of candidate rankings")
    distribution: ConfidenceDistribution = Field(..., description="Confidence score distribution")
    uncertainty: UncertaintyIndicators = Field(..., description="Uncertainty indicators")
    model_accuracy: ModelAccuracySummary = Field(..., description="Model accuracy summary")


class AgreementMetrics(BaseModel):
    """AI vs Human agreement metrics."""

    agreement_rate: float = Field(..., description="Percentage of cases where AI and human agreed (0-1)")
    total_comparisons: int = Field(..., description="Total number of AI-human comparisons")
    ai_correct_count: int = Field(..., description="Number of cases where AI was correct")
    human_override_count: int = Field(..., description="Number of cases where human overrode AI")
    human_override_correct: int = Field(..., description="Number of human overrides that were validated as correct")


class EfficiencyMetrics(BaseModel):
    """Efficiency and time-saving metrics from AI recommendations."""

    avg_time_saved_hours: float = Field(..., description="Average time saved per ranking (hours)")
    total_time_saved_hours: float = Field(..., description="Total time saved by AI recommendations (hours)")
    automation_rate: float = Field(..., description="Percentage of rankings automated without human intervention (0-1)")
    manual_review_rate: float = Field(..., description="Percentage of rankings requiring manual review (0-1)")


class ConfidenceCorrelation(BaseModel):
    """Correlation between AI confidence and actual accuracy."""

    high_confidence_accuracy: float = Field(..., description="Accuracy when AI confidence >= 0.8 (0-1)")
    medium_confidence_accuracy: float = Field(..., description="Accuracy when AI confidence 0.5-0.8 (0-1)")
    low_confidence_accuracy: float = Field(..., description="Accuracy when AI confidence < 0.5 (0-1)")
    correlation_coefficient: float = Field(..., description="Correlation between confidence and accuracy (-1 to 1)")


class AIHumanComparisonResponse(BaseModel):
    """Response model for AI vs human recruiter comparison."""

    agreement: AgreementMetrics = Field(..., description="Agreement metrics between AI and human decisions")
    efficiency: EfficiencyMetrics = Field(..., description="Time-saving and efficiency metrics")
    confidence_correlation: ConfidenceCorrelation = Field(..., description="Correlation between confidence and accuracy")
    recommendations: list[str] = Field(..., description="Recommendations for optimal AI-human collaboration")


class ConfidenceExportRequest(BaseModel):
    """Request model for confidence report data export."""

    format: ConfidenceExportFormat = Field(
        default=ConfidenceExportFormat.JSON,
        description="Export format (json or csv)",
    )
    start_date: Optional[str] = Field(
        None,
        description="Start date for confidence data (ISO 8601 format)",
    )
    end_date: Optional[str] = Field(
        None,
        description="End date for confidence data (ISO 8601 format)",
    )
    vacancy_id: Optional[str] = Field(
        None,
        description="Filter by specific vacancy ID",
    )
    include_metadata: bool = Field(
        default=True,
        description="Whether to include export metadata in the response",
    )


class ConfidenceExportMetadata(BaseModel):
    """Metadata about a confidence report export."""

    export_timestamp: str = Field(
        ...,
        description="ISO 8601 timestamp when export was generated",
    )
    format: str = Field(
        ...,
        description="Export format used (json or csv)",
    )
    start_date: Optional[str] = Field(
        None,
        description="Start date filter applied (ISO 8601 format)",
    )
    end_date: Optional[str] = Field(
        None,
        description="End date filter applied (ISO 8601 format)",
    )
    total_records: int = Field(
        ...,
        description="Total number of records in export",
    )
    filters_applied: Dict[str, Any] = Field(
        ...,
        description="Filters applied to the export",
    )


class ConfidenceExportResponse(BaseModel):
    """Response model for confidence report export."""

    format: str = Field(..., description="Export format used (json or csv)")
    data: Any = Field(..., description="Exported data (dict for JSON, string for CSV)")
    metadata: Optional[ConfidenceExportMetadata] = Field(None, description="Export metadata")


@router.get(
    "/metrics",
    response_model=ConfidenceMetricsResponse,
    tags=["Confidence Dashboard"],
)
async def get_confidence_metrics(
    start_date: Optional[str] = Query(None, description="Start date filter (ISO 8601 format)"),
    end_date: Optional[str] = Query(None, description="End date filter (ISO 8601 format)"),
    vacancy_id: Optional[str] = Query(None, description="Filter by specific vacancy UUID"),
) -> JSONResponse:
    """
    Get aggregated confidence metrics for the transparency dashboard.

    This endpoint provides comprehensive metrics about AI confidence scores across all
    candidate rankings, including:
    - Average, median, min, and max confidence scores
    - Confidence distribution (high/medium/low)
    - Uncertainty indicators for flagging manual review cases
    - Model accuracy summary based on historical performance

    These metrics help recruiters understand when to trust AI recommendations versus
    when to conduct manual review, addressing the "black-box AI" pain point.

    Args:
        start_date: Optional start date for filtering metrics (ISO 8601 format)
        end_date: Optional end date for filtering metrics (ISO 8601 format)
        vacancy_id: Optional filter to show metrics for a specific job vacancy

    Returns:
        JSON response with aggregated confidence metrics

    Raises:
        HTTPException(500): If data retrieval fails

    Examples:
        >>> import requests
        >>> response = requests.get("/api/confidence-dashboard/metrics")
        >>> response.json()
        {
            "average_confidence": 0.76,
            "median_confidence": 0.78,
            "min_confidence": 0.32,
            "max_confidence": 0.96,
            "total_rankings": 1250,
            "distribution": {
                "high_confidence": 875,
                "medium_confidence": 290,
                "low_confidence": 85,
                "high_confidence_percentage": 70.0,
                "medium_confidence_percentage": 23.2,
                "low_confidence_percentage": 6.8
            },
            "uncertainty": {
                "low_confidence_count": 85,
                "high_uncertainty_count": 42,
                "average_confidence_interval_width": 0.18,
                "flagged_for_review": 127
            },
            "model_accuracy": {
                "overall_accuracy": 0.87,
                "precision": 0.85,
                "recall": 0.89,
                "f1_score": 0.87,
                "total_predictions": 1250,
                "feedback_count": 342
            }
        }
    """
    try:
        logger.info(
            f"Fetching confidence metrics - start_date: {start_date}, "
            f"end_date: {end_date}, vacancy_id: {vacancy_id}"
        )

        # For now, return placeholder response
        # Database integration will be added in a later subtask when we have async session setup
        # The data will be aggregated from the CandidateRank table:
        # - prediction_confidence field for average/median/distribution
        # - confidence_interval field for uncertainty indicators
        # - Cross-referenced with RankingFeedback for model accuracy metrics
        response_data = {
            "average_confidence": 0.76,
            "median_confidence": 0.78,
            "min_confidence": 0.32,
            "max_confidence": 0.96,
            "total_rankings": 1250,
            "distribution": {
                "high_confidence": 875,
                "medium_confidence": 290,
                "low_confidence": 85,
                "high_confidence_percentage": 70.0,
                "medium_confidence_percentage": 23.2,
                "low_confidence_percentage": 6.8,
            },
            "uncertainty": {
                "low_confidence_count": 85,
                "high_uncertainty_count": 42,
                "average_confidence_interval_width": 0.18,
                "flagged_for_review": 127,
            },
            "model_accuracy": {
                "overall_accuracy": 0.87,
                "precision": 0.85,
                "recall": 0.89,
                "f1_score": 0.87,
                "total_predictions": 1250,
                "feedback_count": 342,
            },
        }

        logger.info("Confidence metrics retrieved successfully")

        return JSONResponse(
            status_code=status.HTTP_200_OK,
            content=response_data,
        )

    except Exception as e:
        logger.error(f"Error retrieving confidence metrics: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to retrieve confidence metrics: {str(e)}",
        ) from e


@router.get(
    "/ai-human-comparison",
    response_model=AIHumanComparisonResponse,
    tags=["Confidence Dashboard"],
)
async def get_ai_human_comparison(
    start_date: Optional[str] = Query(None, description="Start date filter (ISO 8601 format)"),
    end_date: Optional[str] = Query(None, description="End date filter (ISO 8601 format)"),
    vacancy_id: Optional[str] = Query(None, description="Filter by specific vacancy UUID"),
) -> JSONResponse:
    """
    Get AI vs human recruiter comparison metrics.

    This endpoint provides detailed comparison metrics between AI rankings and human
    recruiter decisions, helping to:
    - Identify when AI recommendations align with human expertise
    - Measure time savings from AI automation
    - Validate the correlation between AI confidence scores and actual accuracy
    - Provide data-driven recommendations for optimal AI-human collaboration

    The comparison is based on:
    - RankingFeedback table: Records recruiter feedback on AI rankings
    - CandidateRank table: AI predictions with confidence scores
    - Learning analytics: Historical accuracy validation

    Args:
        start_date: Optional start date for filtering comparisons (ISO 8601 format)
        end_date: Optional end date for filtering comparisons (ISO 8601 format)
        vacancy_id: Optional filter to compare for a specific job vacancy

    Returns:
        JSON response with AI-human comparison metrics including agreement rates,
        efficiency metrics, confidence correlation, and collaboration recommendations

    Raises:
        HTTPException(500): If data retrieval fails

    Examples:
        >>> import requests
        >>> response = requests.get("/api/confidence-dashboard/ai-human-comparison")
        >>> response.json()
        {
            "agreement": {
                "agreement_rate": 0.82,
                "total_comparisons": 456,
                "ai_correct_count": 374,
                "human_override_count": 82,
                "human_override_correct": 68
            },
            "efficiency": {
                "avg_time_saved_hours": 2.5,
                "total_time_saved_hours": 1140.0,
                "automation_rate": 0.78,
                "manual_review_rate": 0.22
            },
            "confidence_correlation": {
                "high_confidence_accuracy": 0.94,
                "medium_confidence_accuracy": 0.76,
                "low_confidence_accuracy": 0.48,
                "correlation_coefficient": 0.87
            },
            "recommendations": [
                "Trust AI recommendations when confidence >= 0.8 (94% accuracy)",
                "Manual review recommended for confidence < 0.5 (48% accuracy)",
                "AI automation saves average 2.5 hours per ranking",
                "Consider expanding AI automation to medium-confidence cases"
            ]
        }
    """
    try:
        logger.info(
            f"Fetching AI-human comparison - start_date: {start_date}, "
            f"end_date: {end_date}, vacancy_id: {vacancy_id}"
        )

        # For now, return placeholder response
        # Database integration will be added in a later subtask when we have async session setup
        # The data will be aggregated from:
        # - RankingFeedback table: recruiter_ranking vs ai_ranking comparison
        # - CandidateRank table: prediction_confidence correlation with feedback accuracy
        # - Time metrics: Compare manual vs AI-assisted ranking times
        response_data = {
            "agreement": {
                "agreement_rate": 0.82,
                "total_comparisons": 456,
                "ai_correct_count": 374,
                "human_override_count": 82,
                "human_override_correct": 68,
            },
            "efficiency": {
                "avg_time_saved_hours": 2.5,
                "total_time_saved_hours": 1140.0,
                "automation_rate": 0.78,
                "manual_review_rate": 0.22,
            },
            "confidence_correlation": {
                "high_confidence_accuracy": 0.94,
                "medium_confidence_accuracy": 0.76,
                "low_confidence_accuracy": 0.48,
                "correlation_coefficient": 0.87,
            },
            "recommendations": [
                "Trust AI recommendations when confidence >= 0.8 (94% accuracy)",
                "Manual review recommended for confidence < 0.5 (48% accuracy)",
                "AI automation saves average 2.5 hours per ranking",
                "Consider expanding AI automation to medium-confidence cases",
            ],
        }

        logger.info("AI-human comparison metrics retrieved successfully")

        return JSONResponse(
            status_code=status.HTTP_200_OK,
            content=response_data,
        )

    except Exception as e:
        logger.error(f"Error retrieving AI-human comparison: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to retrieve AI-human comparison: {str(e)}",
        ) from e


@router.post(
    "/export",
    response_model=ConfidenceExportResponse,
    tags=["Confidence Dashboard"],
)
async def export_confidence_report(
    request: ConfidenceExportRequest = Body(...),
) -> JSONResponse:
    """
    Export confidence report data in CSV or JSON format.

    This endpoint allows recruiters and managers to export comprehensive confidence
    scoring data for reporting, analysis, and compliance purposes. The export includes:
    - Individual candidate rankings with confidence scores
    - Confidence intervals and uncertainty metrics
    - Model accuracy and performance metrics
    - AI-human comparison data
    - Distribution statistics

    Supports both JSON (for structured data analysis) and CSV (for spreadsheet import)
    formats with optional metadata for audit trails.

    Args:
        request: Export request containing format, date filters, and options

    Returns:
        JSON response containing:
            - format: Export format used ("json" or "csv")
            - data: Exported data (dict for JSON, CSV string for CSV format)
            - metadata: Export metadata with timestamp, filters, and record count

    Raises:
        HTTPException(400): If invalid parameters provided
        HTTPException(500): If export generation fails

    Examples:
        >>> import requests
        >>> response = requests.post("/api/confidence-dashboard/export", json={
        ...     "format": "csv",
        ...     "start_date": "2024-01-01T00:00:00Z",
        ...     "include_metadata": True
        ... })
        >>> response.json()
        {
            "format": "csv",
            "data": "section,metric,value\\nconfidence,average,0.76\\n...",
            "metadata": {
                "export_timestamp": "2024-03-15T10:30:00Z",
                "format": "csv",
                "total_records": 1250,
                "filters_applied": {"vacancy_id": null}
            }
        }
    """
    try:
        logger.info(
            f"Exporting confidence report - format: {request.format}, "
            f"start_date: {request.start_date}, end_date: {request.end_date}, "
            f"vacancy_id: {request.vacancy_id}"
        )

        # For now, collect placeholder data
        # Database integration will be added in a later subtask when we have async session setup
        # The data will be aggregated from:
        # - CandidateRank table: prediction_confidence, confidence_interval
        # - RankingFeedback table: feedback metrics
        # - Cross-referenced for accuracy calculations
        confidence_data = _collect_confidence_data(
            start_date=request.start_date,
            end_date=request.end_date,
            vacancy_id=request.vacancy_id,
        )

        # Format the data based on requested format
        if request.format == ConfidenceExportFormat.CSV:
            formatted_data = _format_as_csv(confidence_data)
        else:  # JSON
            formatted_data = _format_as_json(confidence_data)

        # Calculate total records
        total_records = _count_records(confidence_data)

        # Prepare metadata
        metadata = ConfidenceExportMetadata(
            export_timestamp=datetime.utcnow().isoformat() + "Z",
            format=request.format.value,
            start_date=request.start_date,
            end_date=request.end_date,
            total_records=total_records,
            filters_applied={
                "vacancy_id": request.vacancy_id,
            },
        )

        logger.info(
            f"Confidence export completed: {total_records} records, "
            f"format={request.format.value}"
        )

        result = {
            "format": request.format.value,
            "data": formatted_data,
        }

        if request.include_metadata:
            result["metadata"] = metadata.model_dump()

        return JSONResponse(
            status_code=status.HTTP_200_OK,
            content=result,
        )

    except ValueError as e:
        logger.error(f"Invalid export parameters: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Invalid export parameters: {str(e)}",
        ) from e
    except Exception as e:
        logger.error(f"Error exporting confidence report: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to export confidence report: {str(e)}",
        ) from e


def _collect_confidence_data(
    start_date: Optional[str] = None,
    end_date: Optional[str] = None,
    vacancy_id: Optional[str] = None,
) -> Dict[str, Any]:
    """
    Collect confidence report data for export.

    Args:
        start_date: Optional start date filter (ISO 8601 format)
        end_date: Optional end date filter (ISO 8601 format)
        vacancy_id: Optional vacancy filter

    Returns:
        Dictionary containing confidence data sections
    """
    # Placeholder data structure
    # In production, this will query the database for:
    # - Individual rankings with confidence scores
    # - Aggregated metrics
    # - Distribution data
    # - AI-human comparison data
    return {
        "summary_metrics": {
            "average_confidence": 0.76,
            "median_confidence": 0.78,
            "min_confidence": 0.32,
            "max_confidence": 0.96,
            "total_rankings": 1250,
        },
        "distribution": {
            "high_confidence": 875,
            "medium_confidence": 290,
            "low_confidence": 85,
            "high_confidence_percentage": 70.0,
            "medium_confidence_percentage": 23.2,
            "low_confidence_percentage": 6.8,
        },
        "uncertainty": {
            "low_confidence_count": 85,
            "high_uncertainty_count": 42,
            "average_confidence_interval_width": 0.18,
            "flagged_for_review": 127,
        },
        "model_accuracy": {
            "overall_accuracy": 0.87,
            "precision": 0.85,
            "recall": 0.89,
            "f1_score": 0.87,
            "total_predictions": 1250,
            "feedback_count": 342,
        },
        "ai_human_comparison": {
            "agreement_rate": 0.82,
            "total_comparisons": 456,
            "high_confidence_accuracy": 0.94,
            "medium_confidence_accuracy": 0.76,
            "low_confidence_accuracy": 0.48,
        },
    }


def _format_as_json(confidence_data: Dict[str, Any]) -> Dict[str, Any]:
    """
    Format confidence data as JSON.

    Args:
        confidence_data: Collected confidence data

    Returns:
        JSON-formatted data structure
    """
    return confidence_data


def _format_as_csv(confidence_data: Dict[str, Any]) -> str:
    """
    Format confidence data as CSV.

    Creates a flattened CSV representation of confidence report data.
    Each section is written as separate records with section and metric identifiers.

    Args:
        confidence_data: Collected confidence data

    Returns:
        CSV-formatted string
    """
    output = io.StringIO()
    records = []

    # Flatten summary metrics
    if "summary_metrics" in confidence_data:
        sm = confidence_data["summary_metrics"]
        for metric, value in sm.items():
            records.append({
                "section": "summary_metrics",
                "metric": metric,
                "value": value,
            })

    # Flatten distribution
    if "distribution" in confidence_data:
        dist = confidence_data["distribution"]
        for metric, value in dist.items():
            records.append({
                "section": "distribution",
                "metric": metric,
                "value": value,
            })

    # Flatten uncertainty indicators
    if "uncertainty" in confidence_data:
        unc = confidence_data["uncertainty"]
        for metric, value in unc.items():
            records.append({
                "section": "uncertainty",
                "metric": metric,
                "value": value,
            })

    # Flatten model accuracy
    if "model_accuracy" in confidence_data:
        ma = confidence_data["model_accuracy"]
        for metric, value in ma.items():
            records.append({
                "section": "model_accuracy",
                "metric": metric,
                "value": value,
            })

    # Flatten AI-human comparison
    if "ai_human_comparison" in confidence_data:
        ahc = confidence_data["ai_human_comparison"]
        for metric, value in ahc.items():
            records.append({
                "section": "ai_human_comparison",
                "metric": metric,
                "value": value,
            })

    # Write CSV
    if records:
        writer = csv.DictWriter(output, fieldnames=["section", "metric", "value"])
        writer.writeheader()
        writer.writerows(records)

    return output.getvalue()


def _count_records(confidence_data: Dict[str, Any]) -> int:
    """
    Count total records in confidence data.

    Args:
        confidence_data: Collected confidence data

    Returns:
        Total number of data records
    """
    count = 0

    # Count all key-value pairs across all sections
    for section, data in confidence_data.items():
        if isinstance(data, dict):
            count += len(data)
        elif isinstance(data, list):
            count += len(data)
        else:
            count += 1

    return count

File: frontend/src/api/confidenceDashboard.ts
/**
 * Confidence Dashboard API Client
 *
 * This module provides a convenient interface for the AI Confidence Scoring
 * & Transparency Dashboard, including confidence metrics, model accuracy trends,
 * AI vs human comparisons, and confidence report export functionality.
 *
 * @example
 * ```ts
 * import { confidenceDashboard } from '@/api/confidenceDashboard';
 *
 * // Get confidence metrics
 * const metrics = await confidenceDashboard.getConfidenceMetrics({
 *   start_date: '2024-01-01T00:00:00Z',
 *   end_date: '2024-03-21T00:00:00Z',
 *   vacancy_id: 'vac-123',
 * });
 *
 * // Get model accuracy trend
 * const trend = await confidenceDashboard.getModelAccuracyTrend({
 *   period: '30d',
 *   model_name: 'skill_matching',
 * });
 *
 * // Get AI vs human comparison
 * const comparison = await confidenceDashboard.getAIHumanComparison({
 *   start_date: '2024-01-01T00:00:00Z',
 * });
 *
 * // Export confidence report
 * const report = await confidenceDashboard.exportConfidenceReport({
 *   format: 'csv',
 *   start_date: '2024-01-01T00:00:00Z',
 *   include_metadata: true,
 * });
 * ```
 */

import axios, { AxiosInstance } from 'axios';
import type { ApiError } from '@/types/api';

// ==================== Types ====================

/**
 * Confidence score distribution across different ranges
 */
export interface ConfidenceDistribution {
  high_confidence: number;
  medium_confidence: number;
  low_confidence: number;
  high_confidence_percentage: number;
  medium_confidence_percentage: number;
  low_confidence_percentage: number;
}

/**
 * Indicators for model uncertainty and edge cases
 */
export interface UncertaintyIndicators {
  low_confidence_count: number;
  high_uncertainty_count: number;
  average_confidence_interval_width: number;
  flagged_for_review: number;
}

/**
 * Summary of model accuracy based on historical performance
 */
export interface ModelAccuracySummary {
  overall_accuracy: number;
  precision: number;
  recall: number;
  f1_score: number;
  total_predictions: number;
  feedback_count: number;
}

/**
 * Response model for confidence dashboard metrics
 */
export interface ConfidenceMetricsResponse {
  average_confidence: number;
  median_confidence: number;
  min_confidence: number;
  max_confidence: number;
  total_rankings: number;
  distribution: ConfidenceDistribution;
  uncertainty: UncertaintyIndicators;
  model_accuracy: ModelAccuracySummary;
}

/**
 * Request parameters for confidence metrics
 */
export interface ConfidenceMetricsRequest {
  start_date?: string;
  end_date?: string;
  vacancy_id?: string;
}

/**
 * Single data point in model accuracy trend
 */
export interface ModelAccuracyDataPoint {
  timestamp: string;
  accuracy: number;
  precision: number;
  recall: number;
  f1_score: number;
  sample_size: number;
}

/**
 * Model accuracy trend over time
 */
export interface ModelAccuracyTrendResponse {
  model_name: string;
  period: string;
  data_points: ModelAccuracyDataPoint[];
  summary: {
    avg_accuracy: number;
    avg_f1_score: number;
    trend: string;
    total_samples: number;
  };
}

/**
 * Request parameters for model accuracy trend
 */
export interface ModelAccuracyTrendRequest {
  period?: string;
  model_name?: string;
}

/**
 * AI vs Human agreement metrics
 */
export interface AgreementMetrics {
  agreement_rate: number;
  total_comparisons: number;
  ai_correct_count: number;
  human_override_count: number;
  human_override_correct: number;
}

/**
 * Efficiency and time-saving metrics from AI recommendations
 */
export interface EfficiencyMetrics {
  avg_time_saved_hours: number;
  total_time_saved_hours: number;
  automation_rate: number;
  manual_review_rate: number;
}

/**
 * Correlation between AI confidence and actual accuracy
 */
export interface ConfidenceCorrelation {
  high_confidence_accuracy: number;
  medium_confidence_accuracy: number;
  low_confidence_accuracy: number;
  correlation_coefficient: number;
}

/**
 * Response model for AI vs human recruiter comparison
 */
export interface AIHumanComparisonResponse {
  agreement: AgreementMetrics;
  efficiency: EfficiencyMetrics;
  confidence_correlation: ConfidenceCorrelation;
  recommendations: string[];
}

/**
 * Request parameters for AI vs human comparison
 */
export interface AIHumanComparisonRequest {
  start_date?: string;
  end_date?: string;
  vacancy_id?: string;
}

/**
 * Supported export formats for confidence report data
 */
export enum ConfidenceExportFormat {
  JSON = 'json',
  CSV = 'csv',
}

/**
 * Request model for confidence report data export
 */
export interface ConfidenceExportRequest {
  format?: ConfidenceExportFormat;
  start_date?: string;
  end_date?: string;
  vacancy_id?: string;
  include_metadata?: boolean;
}

/**
 * Metadata about a confidence report export
 */
export interface ConfidenceExportMetadata {
  export_timestamp: string;
  format: string;
  start_date?: string;
  end_date?: string;
  total_records: number;
  filters_applied: {
    vacancy_id?: string;
  };
}

/**
 * Response model for confidence report export
 */
export interface ConfidenceExportResponse {
  format: string;
  data: any;
  metadata?: ConfidenceExportMetadata;
}

// ==================== API Client Class ====================

/**
 * Confidence Dashboard API client
 */
class ConfidenceDashboardClient {
  private client: AxiosInstance;

  constructor(baseURL: string = '/api') {
    this.client = axios.create({
      baseURL,
      headers: {
        'Content-Type': 'application/json',
      },
    });
  }

  /**
   * Get aggregated confidence metrics for the transparency dashboard
   *
   * @param params - Request parameters (optional filters)
   * @returns Confidence metrics response
   * @throws ApiError if request fails
   *
   * @example
   * ```ts
   * const metrics = await confidenceDashboard.getConfidenceMetrics({
   *   start_date: '2024-01-01T00:00:00Z',
   *   vacancy_id: 'vac-123',
   * });
   * console.log(`Average confidence: ${metrics.average_confidence}`);
   * ```
   */
  async getConfidenceMetrics(
    params?: ConfidenceMetricsRequest
  ): Promise<ConfidenceMetricsResponse> {
    try {
      const response = await this.client.get<ConfidenceMetricsResponse>(
        '/confidence-dashboard/metrics',
        { params }
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error)) {
        const apiError: ApiError = {
          detail: error.response?.data?.detail || error.message,
          status: error.response?.status,
        };
        throw apiError;
      }
      throw error;
    }
  }

  /**
   * Get model accuracy trends with historical performance data
   *
   * @param params - Request parameters (period, model_name)
   * @returns Model accuracy trend response
   * @throws ApiError if request fails
   *
   * @example
   * ```ts
   * const trend = await confidenceDashboard.getModelAccuracyTrend({
   *   period: '30d',
   *   model_name: 'skill_matching',
   * });
   * console.log(`Average accuracy: ${trend.summary.avg_accuracy}`);
   * ```
   */
  async getModelAccuracyTrend(
    params?: ModelAccuracyTrendRequest
  ): Promise<ModelAccuracyTrendResponse> {
    try {
      const response = await this.client.get<ModelAccuracyTrendResponse>(
        '/analytics/model-accuracy-trend',
        { params }
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error)) {
        const apiError: ApiError = {
          detail: error.response?.data?.detail || error.message,
          status: error.response?.status,
        };
        throw apiError;
      }
      throw error;
    }
  }

  /**
   * Get AI vs human recruiter comparison metrics
   *
   * @param params - Request parameters (optional filters)
   * @returns AI vs human comparison response
   * @throws ApiError if request fails
   *
   * @example
   * ```ts
   * const comparison = await confidenceDashboard.getAIHumanComparison({
   *   start_date: '2024-01-01T00:00:00Z',
   * });
   * console.log(`Agreement rate: ${comparison.agreement.agreement_rate}`);
   * ```
   */
  async getAIHumanComparison(
    params?: AIHumanComparisonRequest
  ): Promise<AIHumanComparisonResponse> {
    try {
      const response = await this.client.get<AIHumanComparisonResponse>(
        '/confidence-dashboard/ai-human-comparison',
        { params }
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error)) {
        const apiError: ApiError = {
          detail: error.response?.data?.detail || error.message,
          status: error.response?.status,
        };
        throw apiError;
      }
      throw error;
    }
  }

  /**
   * Export confidence report data in CSV or JSON format
   *
   * @param request - Export request with format and filters
   * @returns Exported confidence report
   * @throws ApiError if request fails
   *
   * @example
   * ```ts
   * const report = await confidenceDashboard.exportConfidenceReport({
   *   format: ConfidenceExportFormat.CSV,
   *   start_date: '2024-01-01T00:00:00Z',
   *   include_metadata: true,
   * });
   * console.log(`Exported ${report.metadata?.total_records} records`);
   * ```
   */
  async exportConfidenceReport(
    request: ConfidenceExportRequest
  ): Promise<ConfidenceExportResponse> {
    try {
      const response = await this.client.post<ConfidenceExportResponse>(
        '/confidence-dashboard/export',
        request
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error)) {
        const apiError: ApiError = {
          detail: error.response?.data?.detail || error.message,
          status: error.response?.status,
        };
        throw apiError;
      }
      throw error;
    }
  }
}

// ==================== Export ====================

/**
 * Default confidence dashboard client instance
 *
 * Use this singleton instance for all confidence dashboard API calls.
 */
export const confidenceDashboard = new ConfidenceDashboardClient();

/**
 * Export confidence dashboard client class for custom instances
 */
export default ConfidenceDashboardClient;
